vllm/entrypoints/openai/api_server.py

```diff
@@ -106,6 +108,21 @@ async def build_async_engine_client(args) -> AsyncIterator[AsyncEngineClient]:
 
     # Otherwise, use the multiprocessing AsyncLLMEngine.
     else:
+        if "PROMETHEUS_MULTIPROC_DIR" not in os.environ:
+            # Make TemporaryDirectory for prometheus multiprocessing
+            # Note: global TemporaryDirectory will be automatically
+            #   cleaned up upon exit.
+            global prometheus_multiproc_dir
+            prometheus_multiproc_dir = tempfile.TemporaryDirectory()
+            os.environ[
+                "PROMETHEUS_MULTIPROC_DIR"] = prometheus_multiproc_dir.name
```

Comment:
You really should not do this here. `PROMETHEUS_MULTIPROC_DIR` should be set at the shell level in something like the docker entrypoint. 

`PROMETHEUS_MULTIPROC_DIR` affects things that are operating inside the `prometheus_client` library and so it should be set before that module ever gets loaded. This is why you are having to go through so much effort to lazy load the library in several places.

I am also pretty sure that setting an environment variable in one process does not get reflected into other python processes under this workload. Meaning that each one of them would have their own little temporary direvtory.

Reply:
Additionally, the directory should also get _created_ as part of whatever entrypoint or bootstrapping shell scripts that launch the multiprocess manager.

For my team, we actually set the variable all the way in the Dockerfile so that it can be used to construct the directory at build container image time. Aand persists all the way through to run time so that it can be used by the application / prometheus library.

Reply:
Thanks @thekuffs! In a general case I would agree, but here we have complete control of the back-end process that we are spawning. It's a single process that's started and does not come and go. There may be other processes started transitively but these do not publish/participate in the metrics.

So given that vLLM is a library it's better for this to be self-contained IMO rather than relying on specific external configuration.

> I am also pretty sure that setting an environment variable in one process does not get reflected into other python processes under this workload. Meaning that each one of them would have their own little temporary directory.

The child process inherits the environment from the parent.

Reply:
If there's only one process why does prometheus have to run in multiprocess mode? Is this like a coordinating parent with an API and a single worker process?

Reply:
We have two processes. 

One is the frontend API server (which serves requests and metrics). The other is the backend llm engine (which does work and generates the metrics). We use ZeroMQ to do RPC between the processes.

Introducing this was done in v0.5.4 to improve performance by avoiding contention between the api server and the engine. It gives us 15% speedup. However this broke Prometheus metrics

I’m not an expert in Prometheus, but the approach here seemed the simplest after doing research. But I’m open to other ideas if you have them!

Reply:
I agree it would be better to set the env variable in the shell script but we do not have this luxury as a library that others typically consume via pip install. We have been very careful in the PR about the order of imports and our test strategy was specifically designed to catch any issues with this

Reply:
To isolate the critical model loop. We want to decouple front-end API and processing. GIL makes this necessary :)

Reply:
You're right. I think this is the easiest way to get metrics out of two processes without creating a nightmare. I also didn't realize this was a library, which invalidates everything I have to say about shell scripts.

Thank you for taking the time to explain. I know I am just a stranger barging in with suggestions.

Reply:
No problem! Thanks for reviewing its very helpful to get another set of eyes

Reply:
+1 thanks @thekuffs for the review and input!